In [1]:
import pandas as pd

In [2]:
import statsmodels.api as sm
df = pd.read_csv('Advertising.csv')
x = df [['TV','radio','newspaper']]
y = df['sales']
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

Find a OLS model with intercept on TV and Radio

In [3]:
x = sm.add_constant(x)

In [4]:
x # A new value equaling to 1 has been added as the intercpt value

const     TV  radio  newspaper
0      1.0  230.1   37.8       69.2
1      1.0   44.5   39.3       45.1
2      1.0   17.2   45.9       69.3
3      1.0  151.5   41.3       58.5
4      1.0  180.8   10.8       58.4
..     ...    ...    ...        ...
195    1.0   38.2    3.7       13.8
196    1.0   94.2    4.9        8.1
197    1.0  177.0    9.3        6.4
198    1.0  283.6   42.0       66.2
199    1.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [5]:
model = sm.OLS(y,x).fit()

In [6]:
model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Tue, 09 Apr 2024   Prob (F-statistic):           1.58e-96
Time:                        21:25:43   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### R- squared and Adjusted R- squared value almost tending to 1,meaning the model is pretty good

In [7]:
import matplotlib.pyplot as plt
x.iloc[:,1:].corr()

TV     radio  newspaper
TV         1.000000  0.054809   0.056648
radio      0.054809  1.000000   0.354104
newspaper  0.056648  0.354104   1.000000

Here, the value is almost equal to 0,thus no such multi collinearity found

### Not much relation between the independant features, so no Multi Collinearity issue 

In [8]:
data = pd.read_csv('Salary_Data.csv')

In [9]:
x = data[['YearsExperience','Age']]
y = data['Salary']

In [10]:
x = sm.add_constant(x)

In [11]:
x

const  YearsExperience   Age
0     1.0              1.1  21.0
1     1.0              1.3  21.5
2     1.0              1.5  21.7
3     1.0              2.0  22.0
4     1.0              2.2  22.2
5     1.0              2.9  23.0
6     1.0              3.0  23.0
7     1.0              3.2  23.3
8     1.0              3.2  23.3
9     1.0              3.7  23.6
10    1.0              3.9  23.9
11    1.0              4.0  24.0
12    1.0              4.0  24.0
13    1.0              4.1  24.0
14    1.0              4.5  25.0
15    1.0              4.9  25.0
16    1.0              5.1  26.0
17    1.0              5.3  27.0
18    1.0              5.9  28.0
19    1.0              6.0  29.0
20    1.0              6.8  30.0
21    1.0              7.1  30.0
22    1.0              7.9  31.0
23    1.0              8.2  32.0
24    1.0              8.7  33.0
25    1.0              9.0  34.0
26    1.0              9.5  35.0
27    1.0              9.6  36.0
28    1.0             10.3  37.0
29    1.0             10.5  38.0

In [12]:
model = sm.OLS(y,x).fit()

In [13]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Tue, 09 Apr 2024   Prob (F-statistic):           1.35e-19
Time:                        21:25:58   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Check P-value of Age, if we consider 95% covariance getting covered, the value should be less than 0.05, but it is more, thus there can be some Collinearity and we can find it out using corr()

In [14]:
x.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

Year and age have 98% correlation, thus we can drop one of the features or maybe combine them into one feature and apply OLS model